src/host/agent.js

```diff
@@ -0,0 +1,295 @@
+  }
+
+  async *logs() {
+    yield* this._logs();
+  }
+
+  async start(options) {
+    const process = (this._process = child_process.spawn(
+      path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../bin/agent.js'),
+      agentCliArgsFromOptionsMap({ ...options, protocol: 'shell' }),
+      { stdio: 'pipe' }
+    ));
+    this.exited = new Promise(resolve => {
+      this._process.once('exit', (code, signal) => {
+        resolve({ code, signal });
+      });
+    });
+    const stderrJob = createJob(async function ({ cancelable }) {
+      let carry = '';
+      for await (const buffer of cancelable(
+        iterateEmitter(process.stderr, 'data', 'end', 'error')
+      )) {
+        carry += buffer.toString();
+      }
+      return carry;
+    });
+    this.ready = (async () => {
+      for await (const message of this._logs()) {
+        if (message.text.includes('Start')) {
+          await stderrJob.cancel();
+          return;
+        }
+      }
+      const stderrOutput = await stderrJob.cancel();
+      throw new Error(`Agent shell exited before it was ready.\n${stderrOutput}`);
+    })();
+    await this.ready;
+  }
+
+  async stop() {
+    if (this._process) {
+      this._process.kill();
+    }
+    this._process = null;
+    await this.exited;
+  }
+
+  _results() {
+    return parseJSONChunks(separateJSONChunks(iterateEmitter(this._process.stdout, 'data', 'end')));
+  }
+
+  async *_logs() {
+    for await (const buffer of iterateEmitter(this._process.stderr, 'data', 'end')) {
+      yield { text: buffer.toString().trim() };
+    }
+  }
+}
+
+class AgentAPI {
+  static get protocol() {
+    return 'api';
+  }
+
+  constructor() {
+    this._testEmitter = null;
+    this._logEmitter = null;
+    this._resultEmitter = null;
+    this.exited = null;
+  }
+
+  async run(test) {
+    this._testEmitter.emit('message', test);
+    for await (const result of iterateEmitter(this._resultEmitter, 'result', 'stop')) {
+      return result;
+    }
+    throw new Error('did not receive test result');
+  }
+
+  async *logs() {
+    yield* iterateEmitter(this._logEmitter, 'message', 'stop');
+  }
+
+  async start(options) {
+    const { log, emitter: logEmitter } = createAgentLogger();
+    this._testEmitter = new EventEmitter();
+    this._logEmitter = logEmitter;
+    this._resultEmitter = new EventEmitter();
+    this.exited = agentMain({
+      runner: await createRunner({ baseUrl: options.referenceBaseUrl, log, mock: options.mock }),
+      log,
+      tests: iterateEmitter(this._testEmitter, 'message', 'stop'),
+      reportResult: result => {
+        this._resultEmitter.emit('result', result);
+      },
+    });
+  }
+
+  async stop() {
+    this._testEmitter.emit('stop');
+    this._logEmitter.emit('stop');
+    this._resultEmitter.emit('stop');
+    await this.exited;
+  }
+}
+
+const AGENT_TOOLS = {
+  fork: [AgentFork],
+  shell: [AgentShell],
+  api: [AgentAPI],
+  auto: [AgentFork, AgentShell, AgentAPI],
```

Comment:
Hmm. From host's perspective, fork and shell have the same use case. API has a slightly different use case, trading the ability to recover from agent crashing for performance. Granted `AgentController` doesn't have recovery implemented yet.

Reply:
`auto`'s use case is kind of a small extension, maybe it should be called an experiment, in recovery. When the agent is started if there is an issue, `auto` will try the next in line, until one successfully starts. Previously `auto` was the default, but I changed it to `fork` since including `api` in it would limit the ability to recover